syntax = "proto3";

service ProductDatabaseServer {
  rpc Get(GetRequest) returns (GetResponse) {}
  rpc Insert(InsertRequest) returns (InsertResponse) {}
  rpc Update(UpdateRequest) returns (UpdateResponse) {}
  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
}

message GetRequest {
  string query = 1;
}

message GetResponse {
  repeated Rows rows = 1;
  Issue error = 2;
}

message InsertRequest {
  string query = 1;
}

message InsertResponse {
  int64 insert_id = 1;
  Issue error = 2;
}

message UpdateRequest {
  string query = 1;
}

message UpdateResponse {
  int64 affected_rows = 1;
  Issue error = 2;
}

message DeleteRequest {
  string query = 1;
}

message DeleteResponse {
  int64 affected_rows = 1;
  Issue error = 2;
}

message Rows {
  repeated ColValue values = 1;
}

message ColValue {
  string column_name = 1;
  string column_value = 2;
}

message Issue {
  int32 error_code = 1;
  string error_message = 2;
}